File: examples/claude_example.ts
/**
 * Claude Agent SDK Example - Agentic Learning SDK
 *
 * This example shows how to use the Agentic Learning SDK with Claude Agent SDK.
 * The SDK automatically captures conversations and manages persistent memory.
 *
 * Prerequisites:
 *     npm install @letta/agentic-learning claude-agent-sdk
 *     export ANTHROPIC_API_KEY="your-api-key"
 *
 * Usage:
 *     npx ts-node examples/claude_example.ts
 */

import { ClaudeSDKClient, ClaudeAgentOptions } from 'claude-agent-sdk';
import { AgenticLearning, withLearning } from '../typescript/src';
import { printU, printA, printG, printR, printMessages } from './utils';

// Configure Claude
const apiKey = process.env.ANTHROPIC_API_KEY;
if (!apiKey) {
  console.error('Error: ANTHROPIC_API_KEY environment variable not set');
  console.error("Please run: export ANTHROPIC_API_KEY='your-api-key'");
  process.exit(1);
}

// Create shared client for all examples
const learningClient = new AgenticLearning({ baseUrl: 'http://localhost:8283' });

// Helper function to sleep
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ============================================
// Example 1: Store Memory
// ============================================

async function example1() {
  console.log('='.repeat(60));
  console.log('Example 1: Store Memory');
  console.log('='.repeat(60));
  console.log();

  // Create a Claude SDK client
  const options = new ClaudeAgentOptions();
  let client = new ClaudeSDKClient(options);

  // Use the learning context - this is all you need!
  await withLearning({ agent: 'claude-demo', client: learningClient }, async () => {
    await client.connect();

    printU('My name is Alice.');
    await client.query('My name is Alice.');
    printA('', '', true);

    for await (const message of client.receiveResponse()) {
      if (message.type === 'assistant') {
        for (const block of message.content || []) {
          if (block.type === 'text') {
            process.stdout.write(block.text);
          }
        }
      }
    }
    console.log('\n');

    await sleep(5000); // wait for memory to persist
  });

  // Create a separate session
  client = new ClaudeSDKClient(options);

  await withLearning({ agent: 'claude-demo', client: learningClient }, async () => {
    await client.connect();

    printU("What's my name?");
    await client.query("What's my name?");
    printA('', '', true);

    for await (const message of client.receiveResponse()) {
      if (message.type === 'assistant') {
        for (const block of message.content || []) {
          if (block.type === 'text') {
            process.stdout.write(block.text);
          }
        }
      }
    }
    console.log('\n');

    await client.disconnect();
  });

  printG('✓ Conversation automatically saved to Letta!\n');
}

// ============================================
// Example 2: Streaming (already streaming by nature)
// ============================================

async function example2() {
  console.log('='.repeat(60));
  console.log('Example 2: Streaming');
  console.log('='.repeat(60));
  console.log();

  const options = new ClaudeAgentOptions();
  let client = new ClaudeSDKClient(options);

  await withLearning({ agent: 'claude-demo', client: learningClient }, async () => {
    await client.connect();

    printU(
      'Letta is my favorite context management service. Can you send me a summary about the product Letta (fka MemGPT) offers?'
    );
    await client.query(
      'Letta is my favorite context management service. Can you send me a summary about the product Letta (fka MemGPT) offers?'
    );
    printA('', '', true);

    for await (const message of client.receiveResponse()) {
      if (message.type === 'assistant') {
        for (const block of message.content || []) {
          if (block.type === 'text') {
            process.stdout.write(block.text);
          }
        }
      }
    }
    console.log('\n');

    await sleep(5000); // wait for memory to persist

    printU("What's my favorite context management service?");
    await client.query("What's my favorite context management service?");
    printA('', '', true);

    for await (const message of client.receiveResponse()) {
      if (message.type === 'assistant') {
        for (const block of message.content || []) {
          if (block.type === 'text') {
            process.stdout.write(block.text);
          }
        }
      }
    }
    console.log('\n');

    await client.disconnect();
  });

  printG('✓ Streaming conversation automatically saved to Letta!\n');
}

// ============================================
// Example 3: Capture-Only Mode and Memory Recall
// ============================================

async function example3() {
  console.log('='.repeat(60));
  console.log('Example 3: Capture-Only Mode and Memory Recall');
  console.log('='.repeat(60));
  console.log();

  const options = new ClaudeAgentOptions();
  let client = new ClaudeSDKClient(options);

  await withLearning({
    agent: 'claude-demo',
    captureOnly: true,
    client: learningClient,
  }, async () => {
    await client.connect();

    printU('I am a software engineer.');
    await client.query('I am a software engineer.');
    printA('', '', true);

    for await (const message of client.receiveResponse()) {
      if (message.type === 'assistant') {
        for (const block of message.content || []) {
          if (block.type === 'text') {
            process.stdout.write(block.text);
          }
        }
      }
    }
    console.log('\n');

    await sleep(5000); // Wait for memory to persist

    printR('Testing recall without default memory injection\n');
    printU("What's my professional background?");
    await client.query("What's my professional background?");
    printA('', '', true);

    for await (const message of client.receiveResponse()) {
      if (message.type === 'assistant') {
        for (const block of message.content || []) {
          if (block.type === 'text') {
            process.stdout.write(block.text);
          }
        }
      }
    }
    console.log('\n');

    await client.disconnect();
  });

  printG('Testing memory recall via learning client\n');
  printU("What's my professional background?");
  const searchResults = await learningClient.memory.search(
    'claude-demo',
    "What's my professional background?"
  );
  printMessages(searchResults);

  printG('\nListing stored message history\n');
  const messages = await learningClient.messages.list('claude-demo', { limit: 12 });
  printMessages(messages);

  printG('✓ Memory recall successful!\n');
}

// ============================================
// Run all examples
// ============================================

async function main() {
  await example1();
  await example2();
  await example3();

  console.log('='.repeat(60));
  console.log('All examples complete!');
  console.log('='.repeat(60));
}

main().catch((error) => {
  console.error('Error running examples:', error);
  process.exit(1);
});

File: examples/claude_example.py
"""
Claude Example - Agentic Learning SDK

This example shows how to use the Agentic Learning SDK with Claude Agent SDK.
The SDK automatically captures conversations and manages persistent memory.

Prerequisites:
    pip install agentic-learning
    pip install claude-agent-sdk
    export ANTHROPIC_API_KEY="your-api-key"

Usage:
    python3 claude_example.py
"""

import os
import sys
import asyncio
from claude_agent_sdk import ClaudeSDKClient, ClaudeAgentOptions, AssistantMessage, TextBlock
from agentic_learning import learning_async, AgenticLearning
from utils import print_u, print_a, print_g, print_r, print_messages

# Configure Claude
api_key = os.getenv("ANTHROPIC_API_KEY")
if not api_key:
    print("Error: ANTHROPIC_API_KEY environment variable not set")
    print("Please run: export ANTHROPIC_API_KEY='your-api-key'")
    sys.exit(1)

# Create sync client for memory injection (works with async contexts)
learning_client = AgenticLearning(base_url="http://localhost:8283")


# ============================================
# Example 1: Store Memory
# ============================================

async def example_1():
    print("=" * 60)
    print("Example 1: Store Memory")
    print("=" * 60)
    print()

    # Create a Claude SDK client
    options = ClaudeAgentOptions()
    client = ClaudeSDKClient(options)

    # Use the learning context - this is all you need!
    async with learning_async(agent="claude-demo", client=learning_client):
        await client.connect()

        print_u("My name is Alice.")
        await client.query("My name is Alice.")
        print_a("", end="", flush=True)
        async for message in client.receive_response():
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        print(block.text, end="", flush=True)
        print("\n")

        await client.disconnect()

    # Brief wait to ensure message save completes
    await asyncio.sleep(10)

    # Create a new session - memory will be injected at initialization
    client = ClaudeSDKClient(options)

    async with learning_async(agent="claude-demo", client=learning_client):
        await client.connect()

        print_u("What's my name?")
        await client.query("What's my name?")
        print_a("", end="", flush=True)
        async for message in client.receive_response():
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        print(block.text, end="", flush=True)
        print("\n")

        await client.disconnect()

    print_g("✓ Conversation automatically saved to Letta!\n")


# ============================================
# Example 2: Streaming (already streaming by nature)
# ============================================

async def example_2():
    print("=" * 60)
    print("Example 2: Streaming")
    print("=" * 60)
    print()

    options = ClaudeAgentOptions()
    client = ClaudeSDKClient(options)

    async with learning_async(agent="claude-demo", client=learning_client):
        await client.connect()

        print_u("Letta is my favorite context management service. Can you send me a summary about the product Letta (fka MemGPT) offers?")
        await client.query("Letta is my favorite context management service. Can you send me a summary about the product Letta (fka MemGPT) offers?")
        print_a("", end="", flush=True)

        async for message in client.receive_response():
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        print(block.text, end="", flush=True)

        print("\n")

        await client.disconnect()

    # Brief wait to ensure message save completes
    await asyncio.sleep(10)

    # Create a new session - memory will be injected at initialization
    client = ClaudeSDKClient(options)

    async with learning_async(agent="claude-demo", client=learning_client):
        await client.connect()

        print_u("What's my favorite context management service?")
        await client.query("What's my favorite context management service?")
        print_a("", end="", flush=True)

        async for message in client.receive_response():
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        print(block.text, end="", flush=True)

        print("\n")

        await client.disconnect()

    print_g("✓ Streaming conversation automatically saved to Letta!\n")


# ===============================================
# Example 3: Capture-Only Mode and Memory Recall
# ===============================================

async def example_3():
    print("=" * 60)
    print("Example 3: Capture-Only Mode and Memory Recall")
    print("=" * 60)
    print()

    options = ClaudeAgentOptions()
    client = ClaudeSDKClient(options)

    async with learning_async(agent="claude-demo", capture_only=True, client=learning_client):
        await client.connect()

        print_u("I am a software engineer.")
        await client.query("I am a software engineer.")
        print_a("", end="", flush=True)
        async for message in client.receive_response():
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        print(block.text, end="", flush=True)
        print("\n")

        await client.disconnect()

    # Brief wait to ensure message save completes
    await asyncio.sleep(10)

    # Create a separate session
    client = ClaudeSDKClient(options)

    async with learning_async(agent="claude-demo", capture_only=True, client=learning_client):
        await client.connect()

        print_r("Testing recall without memory injection (capture_only=True)\n")
        print_u("What's my professional background?")
        await client.query("What's my professional background?")
        print_a("", end="", flush=True)
        async for message in client.receive_response():
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        print(block.text, end="", flush=True)
        print("\n")

        await client.disconnect()

    print_g("Testing memory recall via learning client\n")
    print_u("What's my professional background?")
    messages = learning_client.memory.search("claude-demo", "What's my professional background?")
    print_messages(messages)

    print_g("✓ Memory recall successful!\n")

    print_g("\nListing stored message history\n")
    messages = learning_client.messages.list("claude-demo", limit=12)
    print_messages(messages)

    print_g("✓ Message recall successful!\n")


async def main():
    await example_1()
    await example_2()
    await example_3()

    print("=" * 60)
    print("All examples complete!")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())
